docs/design/1161-standardize-stdlib.md

```diff
@@ -0,0 +1,247 @@
+    * `string.join`
+    * `string.replace`
+    * `string.split`
+    * `string.to_lower`
+    * `string.to_upper`
+    * `string.trim`
+    * `string.trim_prefix`
+    * `string.trim_space`
+    * `string.trim_suffix`
+    * `string.contains` (**new**; check if a string contains a substring)
+
+## Pros and cons
+
+Pros:
+
+* Aligns with how constants and component names are namespaced.
+* Makes it clear which identifiers are standardized by looking for the
+  dot-separated namespace.
+* Namespaces enable having two functions with the same name in two different
+  namespaces (`path.join` and `string.join`).
+
+Cons:
+
+* Deprecating 17 of the 19 identifiers may be seen as aggressive, even if they
+  are never removed.
+* Using the standard library becomes more verbose
+    * However, some of the verbosity may be bought back in the future if
+      [aliases](https://github.com/grafana/alloy/issues/154) are introduced.
+* Namespaces in the standard library must not collide with a component
+  namespace.
+
+## Alternative solutions
+
+### Separate namespace and identifier with underscore
+
+Rather than separating the namespace and identifier using a dot
+(`string.join`), an underscore could be used instead (`string_join`).
+
+Pros:
+
+* Clearer when something is in the standard library compared to being a
+  reference to a component.
+* Not possible for a standard library namespace to collide with a component
+  namespace.
+
+Cons:
+
+* Inconsistent with `constants` which uses dot-separation for retrieving a
+  specific constant.
+    * This could be mitigated by changing `constants.os` to `constants_os` or
+      similar.
+
+### Do not standardize existing identifiers
+
+Alternatively, a pattern for standardizing new identifiers could be
+established, while existing identifiers are left untouched.
+
+Pros:
+
+* The majority of the existing standard library remains valid.
+
+Cons:
+
+* Until the number of new identifiers is far more than the number of current
+  identifiers, the standard library will appear very inconsistent, with a heavy
+  mix of namespaced and non-namespaced identifiers.
+* An existing identifier may belong in a namespace alongside new identifiers,
+  but is left out for backwards compatibility, causing inconsistencies.
+
+### Do not establish a standardization for identifiers
+
+Alternatively, we can omit standardizing identifiers at all, and continue to
+let the standard library grow organically.
+
+Pros:
+
+* Less work for contributors :)
+
+Cons:
+
+* It will be harder to understand and navigate the standard library as the
+  number of identifiers grows.
+
+## Compatibility
+
+This proposal is properly backwards compatible, with the old identifiers being
+marked deprecated in favour of their namespaced equivalents (where a namespace
+was introduced).
+
+It is possible that any namespaces introduced in this proposal may collide with
+an existing custom component. Custom components shadow the stdlib, so this is
```

Comment:
Ha! Good call out. It's nice that custom components take precedence.

Reply:
Credit to @wildum :stuck_out_tongue: